### PR TITLE
[TD] avoid memory leaks by using shared_ptr

### PR DESCRIPTION
**Pull request by WandererFan**
TD geometry objects are sometimes double deleted.  This change uses shared_ptr instead of raw pointers to manage
deletions.

An expansion of the patch discussed here: [forum topic](https://www.forum.freecadweb.org/viewtopic.php?p=535784#p535784l) to BaseGeometry and derived classes.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 1]
```diff
@@ -344,14 +344,14 @@ CosmeticEdge::CosmeticEdge()
 {
 //    Base::Console().Message("CE::CE()\n");
     permaRadius = 0.0;
-    m_geometry = new TechDraw::BaseGeom();
+    m_geometry = std::make_shared<TechDraw::BaseGeom> (BaseGeom());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:347]: The call of BaseGeom() inside the parentheses can be dropped. What it does is create an instance on the stack and afterward it invokes the default copy constructor. So, this should be sufficient: `m_geometry = std::make_shared<TechDraw::BaseGeom> ();`

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 2]
```diff
@@ -410,9 +410,10 @@ CosmeticEdge::CosmeticEdge(TechDraw::BaseGeom* g)
 
 CosmeticEdge::~CosmeticEdge(void)
 {
-    if (m_geometry != nullptr) {
-        delete m_geometry;
-    }
+    //shared pointer will delete m_geometry when ref count goes to zero.
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:413]: The commented code should be removed but if you like you can keep the comment.

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 3]
```diff
@@ -509,25 +510,25 @@ void CosmeticEdge::Restore(Base::XMLReader &reader)
     TechDraw::GeomType gType = (TechDraw::GeomType)reader.getAttributeAsInteger("value");
 
     if (gType == TechDraw::GeomType::GENERIC) {
-        TechDraw::Generic* gen = new TechDraw::Generic();
+        TechDraw::GenericPtr gen = std::make_shared<TechDraw::Generic> (TechDraw::Generic());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:513]: Drop the call of TechDraw::Generic()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 4]
```diff
@@ -509,25 +510,25 @@ void CosmeticEdge::Restore(Base::XMLReader &reader)
...
         gen->Restore(reader);
         gen->occEdge = GeometryUtils::edgeFromGeneric(gen);
-        m_geometry = (TechDraw::BaseGeom*) gen;
+        m_geometry = (TechDraw::BaseGeomPtr) gen;
         permaStart = gen->getStartPoint();
         permaEnd   = gen->getEndPoint();
     } else if (gType == TechDraw::GeomType::CIRCLE) {
-        TechDraw::Circle* circ = new TechDraw::Circle();
+        TechDraw::CirclePtr circ = std::make_shared<TechDraw::Circle> (TechDraw::Circle());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:520]: Drop the call of TechDraw::Circle()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 5]
```diff
@@ -509,25 +510,25 @@ void CosmeticEdge::Restore(Base::XMLReader &reader)
...
         circ->Restore(reader);
         circ->occEdge = GeometryUtils::edgeFromCircle(circ);
-        m_geometry = (TechDraw::BaseGeom*) circ;
+        m_geometry = (TechDraw::BaseGeomPtr) circ;
         permaRadius = circ->radius;
         permaStart  = circ->center;
         permaEnd    = circ->center;
     } else if (gType == TechDraw::GeomType::ARCOFCIRCLE) {
-        TechDraw::AOC* aoc = new TechDraw::AOC();
+        TechDraw::AOCPtr aoc = std::make_shared<TechDraw::AOC> (TechDraw::AOC());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:528]: Drop the call of TechDraw::AOC()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 6]
```diff
@@ -614,7 +615,7 @@ CenterLine::CenterLine(void)
     m_type = CLTYPE::FACE;
     m_flip2Line = false;
 
-    m_geometry = new TechDraw::BaseGeom();
+    m_geometry = std::make_shared<TechDraw::BaseGeom> (TechDraw::BaseGeom());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:618]: Drop the call of TechDraw::BaseGeom()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 7]
```diff
@@ -1385,20 +1386,20 @@ void CenterLine::Restore(Base::XMLReader &reader)
     reader.readElement("GeometryType");
     TechDraw::GeomType gType = (TechDraw::GeomType)reader.getAttributeAsInteger("value");
     if (gType == TechDraw::GeomType::GENERIC) {
-        TechDraw::Generic* gen = new TechDraw::Generic();
+        TechDraw::GenericPtr gen = std::make_shared<TechDraw::Generic> (TechDraw::Generic());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:1389]: Drop the call of TechDraw::Generic()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 8]
```diff
@@ -1385,20 +1386,20 @@ void CenterLine::Restore(Base::XMLReader &reader)
...
         gen->Restore(reader);
         gen->occEdge = GeometryUtils::edgeFromGeneric(gen);
-        m_geometry = (TechDraw::BaseGeom*) gen;
+        m_geometry = (TechDraw::BaseGeomPtr) gen;
     } else if (gType == TechDraw::GeomType::CIRCLE) {
-        TechDraw::Circle* circ = new TechDraw::Circle();
+        TechDraw::CirclePtr circ = std::make_shared<TechDraw::Circle> (TechDraw::Circle());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:1394]: Drop the call of TechDraw::Circle()

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 9]
```diff
@@ -1385,20 +1386,20 @@ void CenterLine::Restore(Base::XMLReader &reader)
     reader.readElement("GeometryType");
     TechDraw::GeomType gType = (TechDraw::GeomType)reader.getAttributeAsInteger("value");
     if (gType == TechDraw::GeomType::GENERIC) {
-        TechDraw::Generic* gen = new TechDraw::Generic();
+        TechDraw::GenericPtr gen = std::make_shared<TechDraw::Generic> (TechDraw::Generic());
         gen->Restore(reader);
         gen->occEdge = GeometryUtils::edgeFromGeneric(gen);
-        m_geometry = (TechDraw::BaseGeom*) gen;
+        m_geometry = (TechDraw::BaseGeomPtr) gen;
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:1392]: The C-style cast can be dropped

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 10]
```diff
@@ -1385,20 +1386,20 @@ void CenterLine::Restore(Base::XMLReader &reader)
...
     } else if (gType == TechDraw::GeomType::CIRCLE) {
-        TechDraw::Circle* circ = new TechDraw::Circle();
+        TechDraw::CirclePtr circ = std::make_shared<TechDraw::Circle> (TechDraw::Circle());
         circ->Restore(reader);
         circ->occEdge = GeometryUtils::edgeFromCircle(circ);
-        m_geometry = (TechDraw::BaseGeom*) circ;
+        m_geometry = (TechDraw::BaseGeomPtr) circ;
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:1397]: The C-style cast can be dropped.

File: src/Mod/TechDraw/App/Cosmetic.cpp [hunk 11]
```diff
@@ -1385,20 +1386,20 @@ void CenterLine::Restore(Base::XMLReader &reader)
...
     } else if (gType == TechDraw::GeomType::ARCOFCIRCLE) {
-        TechDraw::AOC* aoc = new TechDraw::AOC();
+        TechDraw::AOCPtr aoc = std::make_shared<TechDraw::AOC> (TechDraw::AOC());
         aoc->Restore(reader);
         aoc->occEdge = GeometryUtils::edgeFromCircleArc(aoc);
-        m_geometry = (TechDraw::BaseGeom*) aoc;
+        m_geometry = (TechDraw::BaseGeomPtr) aoc;
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Cosmetic.cpp:1402]: The C-style cast can be dropped

File: src/Mod/TechDraw/App/CosmeticEdgePyImp.cpp [hunk 12]
```diff
@@ -316,9 +316,9 @@ void CosmeticEdgePy::setRadius(Py::Object arg)
     }
 
     getCosmeticEdgePtr()->permaRadius = r;
-    auto oldGeom = getCosmeticEdgePtr()->m_geometry;
-    getCosmeticEdgePtr()->m_geometry = new TechDraw::Circle(getCosmeticEdgePtr()->permaStart, r);
-    delete oldGeom;
+//    auto oldGeom = getCosmeticEdgePtr()->m_geometry;
+    getCosmeticEdgePtr()->m_geometry = std::make_shared<TechDraw::Circle> (TechDraw::Circle(getCosmeticEdgePtr()->permaStart, r));
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/CosmeticEdgePyImp.cpp:320]: Use `getCosmeticEdgePtr()->m_geometry = std::make_shared<TechDraw::Circle> (getCosmeticEdgePtr()->permaStart, r);`

File: src/Mod/TechDraw/App/CosmeticEdgePyImp.cpp [hunk 13]
```diff
@@ -361,16 +361,16 @@ void CosmeticEdgePy::setCenter(Py::Object arg)
...
     if (oldCircle == nullptr) {
         throw Py::TypeError("Edge geometry is not a circle");
     }
 
     getCosmeticEdgePtr()->permaStart = pNew;
     getCosmeticEdgePtr()->permaEnd = pNew;
     getCosmeticEdgePtr()->permaRadius = oldCircle->radius;
-    getCosmeticEdgePtr()->m_geometry = new TechDraw::Circle(getCosmeticEdgePtr()->permaStart, oldCircle->radius);
-    delete oldGeom;
+    getCosmeticEdgePtr()->m_geometry = std::make_shared<TechDraw::Circle> (TechDraw::Circle(getCosmeticEdgePtr()->permaStart, oldCircle->radius));
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/CosmeticEdgePyImp.cpp:372]: Drop the call of TechDraw::Circle() as above.

File: src/Mod/TechDraw/App/DrawParametricTemplate.cpp [hunk 14]
```diff
@@ -128,10 +129,11 @@ int DrawParametricTemplate::drawLine(double x1, double y1, double x2, double y2)
 
 int DrawParametricTemplate::clearGeometry()
 {
-    for(std::vector<TechDraw::BaseGeom *>::iterator it = geom.begin(); it != geom.end(); ++it) {
-        delete *it;
-        *it = 0;
-    }
+    //smart pointer will delete old geoms when ref count goes to zero?
+//    for(std::vector<TechDraw::BaseGeomPtr>::iterator it = geom.begin(); it != geom.end(); ++it) {
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/DrawParametricTemplate.cpp:133]: Remove the commented code.

File: src/Mod/TechDraw/App/DrawViewPart.cpp [hunk 15]
```diff
@@ -863,8 +863,8 @@ BaseGeom* DrawViewPart::projectEdge(const TopoDS_Edge& e) const
     projector.Build();
     TopoDS_Shape s = projector.Projection();
 //    Base::Console().Message("DVP::projectEdge - s.IsNull: %d\n", s.IsNull());
-//    BaseGeom* result = BaseGeom::baseFactory(pe);
-    BaseGeom* result = nullptr;
+//    BaseGeomPtr result = BaseGeom::baseFactory(pe);
+    BaseGeomPtr result = nullptr;
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/DrawViewPart.cpp:867]: Assigning nullptr is OK but not required.

File: src/Mod/TechDraw/App/DrawViewPartPyImp.cpp [hunk 16]
```diff
@@ -435,7 +435,7 @@ PyObject* DrawViewPartPy::makeCosmeticCircleArc(PyObject *args)
     //from here on is almost duplicate of makeCosmeticCircle
     DrawViewPart* dvp = getDrawViewPartPtr();
     Base::Vector3d pnt1 = DrawUtil::invertY(static_cast<Base::VectorPy*>(pPnt1)->value());
-    TechDraw::BaseGeom* bg = new TechDraw::AOC(pnt1, radius, angle1, angle2);
+    TechDraw::BaseGeomPtr bg = std::make_shared<TechDraw::AOC> (TechDraw::AOC(pnt1, radius, angle1, angle2));
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/DrawViewPartPyImp.cpp:438]: The call of TechDraw::AOC() can be dropped.

File: src/Mod/TechDraw/App/Geometry.cpp [hunk 17]
```diff
@@ -203,7 +201,7 @@ BaseGeom* BaseGeom::copy()
             result->cosmeticTag = cosmeticTag;
         }
     } else {
-        result = new BaseGeom();
+        result = std::make_shared<BaseGeom>(BaseGeom());
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Geometry.cpp:204]: The call of BaseGeom() can be dropped.

File: src/Mod/TechDraw/App/Geometry.cpp [hunk 18]
```diff
@@ -449,7 +447,7 @@ bool BaseGeom::closed(void)
 
 
 //! Convert 1 OCC edge into 1 BaseGeom (static factory method)
-BaseGeom* BaseGeom::baseFactory(TopoDS_Edge edge)
+BaseGeomPtr BaseGeom::baseFactory(TopoDS_Edge edge)
 {
     std::unique_ptr<BaseGeom> result;
```
Comment (wwmayer) [inline src/Mod/TechDraw/App/Geometry.cpp:452]: BaseGeomPtr is declared as std::shared_ptr so result should be a shared_ptr too, although a shared_ptr accepts a unique_ptr.

**Review by wwmayer — CHANGES_REQUESTED** [review]
Please have a look at all my comments.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5323">pipeline status</a> for feature branch [PR_5323](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5323). Pipeline [440637543](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/440637543) was triggered at [60c41f9](https://github.com/FreeCAD/FreeCAD/pull/5323/commits/60c41f9182bfad172294add7235f8b86a5751809). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by WandererFan** [comment]
How are we handling commit history for changes due to code review?  I would say 1 clean commit with the original changes + review changes squashed together, but I'll go with our standard.

**Comment by donovaly** [comment]
> How are we handling commit history for changes due to code review?

There are no defined rules. Common is that the user makes the requested change, rebase and force-push this. The history of a PR is not of interest.
Personally I try to have only one commit per PR, except to keep the overview or for technical reasons.

**Comment by WandererFan** [comment]
> Please have a look at all my comments.

Excellent review!  Thank you very much.

**Comment by luzpaz** [comment]
Addendum: try to preprend `[TD]` (or whatever workbench you're touching) to the commit summary names. That way we can easily parse the git history

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5323">pipeline status</a> for feature branch [PR_5323](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5323). Pipeline [443596420](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/443596420) was triggered at [ec24725](https://github.com/FreeCAD/FreeCAD/pull/5323/commits/ec2472569fee0f14039651e1fbcc050cc87ae687). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Pull request merged**